### PR TITLE
Allow disabling of dependency installation in setup.py

### LINKED ISSUE #1128: Allow disabling of pip installations in setup.py
### What is the expected behavior?

Currently setup.py is quite aggressive in installing dependencies (see https://github.com/Qiskit/qiskit-aer/blob/master/setup.py#L25). This leads to unexpected behaviour on HPC systems where Qiskit is being installed as an environment module (i.e., self-contained) as it installs these dependencies for the current `python`, which in our case is in another environment module. This makes our Python installation non-reproducible, and the Qiskit module broken if another Python environment module is used. It should be possible to disable this behaviour since it can have severe unintended consequences.

### PR DESCRIPTION
**Pull request by vvilpas**
### Summary

- Some Python dependencies are installed outside regular `setuptools` installation because are required
  by compilation of C++ code. Now this installation can be disabled setting the environment variable
  _DISABLE_DEPENDENCY_INSTALL to 1 or "On".

- This commit also adds version checks on those packages when `setup.py` checks if they are already available.

### Details and comments

Fixes #1128

### CODE REVIEW

**Review by mtreinish — APPROVED** [review]
This LGTM, it might be worth pointing out that newer pip versions shouldn't need this because of the pyproject.toml build requirements outlining these. But it comes up enough that people are calling setup.py directly or using an older pip version that this doesn't work.

My only hesitation here is the additional use of distutils which will be deprecated in python 3.10: https://www.python.org/dev/peps/pep-0632/ they specifically call out strtobool as something we need to implement ourselves (https://www.python.org/dev/peps/pep-0632/#migration-advice ). Maybe for future proofing we do that in a follow up and write our own strtobool

**Comment by vvilpas** [comment]
I've added our won `strtobool` function, mimicking  `distutils` one.

File: setup.py [hunk 1]
```diff
@@ -3,36 +3,56 @@
...
-
-try:
-    import pybind11
-except ImportError:
-    subprocess.call([sys.executable, '-m', 'pip', 'install', 'pybind11>=2.6'])
+        mod = importlib.import_module(to_import)
+        mod_ver = parse_version(mod.__version__)
+        if ((min_version and mod_ver < parse_version(min_version))
+                or (max_version and mod_ver >= parse_version(max_version))):
+            raise RuntimeError('{} {} is installed but required version is {}.'.
```
Comment (chriseclectic) [inline setup.py:40]:
```suggestion
            raise RuntimeError(f'{package_name} {mod_ver} is installed but required version is {install_ver}.'.
```

File: setup.py [hunk 2]
```diff
@@ -3,36 +3,56 @@
...
+                               format(to_install, mod_ver, to_install_ver))
+
+    except ImportError as err:
+        if _DISABLE_DEPENDENCY_INSTALL:
+            raise ImportError(str(err) +
+                              "\n{} is a required dependency. Please provide it and repeat install"
```
Comment (chriseclectic) [inline setup.py:46]:
```suggestion
                              f"\n{package_name} is a required dependency. Please provide it and repeat install"
```

**Review by chriseclectic — COMMENTED** [review]
Some suggestions for making the code a little easier to read

**Review by chriseclectic — APPROVED** [review]

**Pull request merged**